govcd/user.go

```diff
@@ -0,0 +1,456 @@
+/*
+ * Copyright 2019 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"github.com/vmware/go-vcloud-director/v2/util"
+)
+
+// Definition of an OrgUser
+type OrgUser struct {
+	User     *types.User
+	client   *Client
+	AdminOrg *AdminOrg // needed to be able to update, as the list of roles is found in the Org
+}
+
+// Simplified structure to insert or modify an organization user
+type OrgUserConfiguration struct {
+	Name            string // Mandatory
+	Password        string // Mandatory
+	RoleName        string // Mandatory
+	ProviderType    string // Optional: defaults to "INTEGRATED"
+	IsEnabled       bool   // Optional: defaults to false
+	IsLocked        bool   // Only used for updates
+	DeployedVmQuota int    // Optional: 0 means "unlimited"
+	StoredVmQuota   int    // Optional: 0 means "unlimited"
+	FullName        string
+	Description     string
+	EmailAddress    string
+	Telephone       string
+}
+
+const (
+	// Common role names and provider types are kept here to reduce hard-coded text and prevent mistakes
+	// Roles that are added to the organization need to be entered as free text
+
+	OrgUserRoleOrganizationAdministrator = "Organization Administrator"
+	OrgUserRoleCatalogAuthor             = "Catalog Author"
+	OrgUserRoleVappAuthor                = "vApp Author"
+	OrgUserRoleVappUser                  = "vApp User"
+	OrgUserRoleConsoleAccessOnly         = "Console Access Only"
+	OrgUserRoleDeferToIdentityProvider   = "Defer to Identity Provider"
+
+	// Allowed values for provider types
+	OrgUserProviderIntegrated = "INTEGRATED" // The user is created locally or imported from LDAP
+	OrgUserProviderSAML       = "SAML"       // The user is imported from a SAML identity provider.
+	OrgUserProviderOAUTH      = "OAUTH"      // The user is imported from an OAUTH identity provider
+)
+
+// Used to check the validity of provider type on creation
+var OrgUserProviderTypes = []string{
+	OrgUserProviderIntegrated,
+	OrgUserProviderSAML,
+	OrgUserProviderOAUTH,
+}
+
+// NewUser creates an empty user
+func NewUser(cli *Client, org *AdminOrg) *OrgUser {
+	return &OrgUser{
+		User:     new(types.User),
+		client:   cli,
+		AdminOrg: org,
+	}
+}
+
+// GetUserByNameOrId retrieves an user within an admin organization
+// by either name or ID
+// Returns a valid user if it exists. If it doesn't, returns nil and ErrorEntityNotFound
+func (adminOrg *AdminOrg) GetUserByNameOrId(identifier string, willRefresh bool) (*OrgUser, error) {
+	if willRefresh {
```

Comment:
Like other functions where we search the parent to find the entity, we need to refresh the entity to make sure we have a fresh image. However, since the refresh is costly, if we know that we want to get what is already there (such as when we want to get all the current users, we don't need to refresh every time) we set `willRefhesh` to `false`.

Reply:
added comment

Reply:
Sorry, I meant 'willRefresh -> 'refresh' or something else

Reply:
"refresh" can be either a verb or a noun, while `willRefresh` shows that I want it to do something.

Reply:
@lvirbalas @Didainius what is your opinion?

Reply:
I can add one more  `withRefresh` :) but it won't change too much so it can be even as it is for me.

Reply:
I have a pretty strong opinion on this one. It should be just `refresh` and *we should use the same approach in other places if/when we need it*.